examples/transport/readme.md

```diff
@@ -0,0 +1,199 @@
+# Implementing a new libp2p transport
+
+Libp2p is the modular networking stack used by ipfs. It is designed in a way that anyone should be able to make it work with little effort over any given transport. In this example, i'm going to go through how to make libp2p (and consequently ipfs) work over websockets. 
+
+To implement a transport on the go side of libp2p, all you need to do is satisfy the `Transport` interface in [go-libp2p-transport](https://github.com/ipfs/go-libp2p-transport). To do this, I'm going to use an existing websockets library from [golang.org/x/net/websocket](https://golang.org/x/net/websocket) and create my own package `ws-transport`.
+
+## Multiaddr Handling
+The first thing we're going to need is a multiaddr parser to be able to reference our websocket addresses. 
+
+Lets create a multiaddr protocol for this:
+```go
+var WsProtocol = ma.Protocol{
+	Code:  477,
+	Name:  "ws",
+	VCode: ma.CodeToVarint(477),
+}
+```
+
+Here we select `ws` as the tag for websockets, and `477` as the binary code for websockets.  We also need to implement a few extra things as shims so we can reuse a lot of the code already written in [go-multiaddr-net](https://github.com/jbenet/go-multiaddr-net).
+
+Next, a `multiaddr-net` shim:
+
+```go
+var WsCodec = &manet.NetCodec{
+	NetAddrNetworks:  []string{"websocket"},
+	ProtocolName:     "ws",
+	ConvertMultiaddr: ConvertWebsocketMultiaddrToNetAddr,
+	ParseNetAddr:     ParseWebsocketNetAddr,
+}
+```
+
+This defines a codec to allow libp2p to translate conventional address types to multiaddr types. The two functions referenced there implement those translations, and you can check out the code [here](https://github.com/whyrusleeping/ws-transport/blob/master/websocket.go#L43).
+
+Now, a formatter object to parse and validate websocket multiaddrs:
+
+```go
+var WsFmt = mafmt.And(mafmt.TCP, mafmt.Base(WsProtocol.Code))
+```
+
+The above statement declares a websocket address as a `TCP` address with a `/ws/` protocol tag at the end, for example: `/ip4/1.2.3.4/tcp/8005/ws`
+
+Now that we have those, we should register them with the multiaddr library in our package initialization function:
+
+```go
+func init() {
+	err := ma.AddProtocol(WsProtocol)
+	if err != nil {
+		log.Fatalf("error registering websocket protocol: %s", err)
+	}
+
+	manet.RegisterNetCodec(WsCodec)
+}
+```
+
+## The Transport
+
+Now that all that nonsense is out of the way, let's get to actually implementing the Transport. To do this, we need four main types, the `Transport`, the `Dialer`, the `Listener` and a `Conn` type.
```

Comment:
`Now that all that nonsense is out of the way` Let's rephrase this or just remove it :)